berkdb/mp/mp_bh.c

```diff
@@ -1173,41 +1142,29 @@ __dir_pg(dbmfp, pgno, buf, is_pgin)
 	if (gbl_bb_berkdb_enable_memp_pg_timing)
 		start_time_us = bb_berkdb_fasttime();
 
-	MUTEX_THREAD_LOCK(dbenv, dbmp->mutexp);
-
 	ftype = mfp->ftype;
-	for (mpreg = LIST_FIRST(&dbmp->dbregq);
-	    mpreg != NULL; mpreg = LIST_NEXT(mpreg, q)) {
-		if (ftype != mpreg->ftype)
-			continue;
-		if (mfp->pgcookie_len == 0)
-			dbtp = NULL;
-		else {
-			dbt.size = mfp->pgcookie_len;
-			dbt.data = R_ADDR(dbmp->reginfo, mfp->pgcookie_off);
-			dbtp = &dbt;
-		}
-		MUTEX_THREAD_UNLOCK(dbenv, dbmp->mutexp);
-
-		if (is_pgin) {
-			if (mpreg->pgin != NULL &&
-			    (ret = mpreg->pgin(dbenv, pgno, buf, dbtp)) != 0)
-				goto err;
-		} else
-		    if (mpreg->pgout != NULL &&
-		    (ret = mpreg->pgout(dbenv, pgno, buf, dbtp)) != 0)
-			goto err;
-		break;
-	}
-
-	if (mpreg == NULL)
-		MUTEX_THREAD_UNLOCK(dbenv, dbmp->mutexp);
+    if (mfp->pgcookie_len == 0)
+        dbtp = NULL;
+    else {
+        dbt.size = mfp->pgcookie_len;
+        dbt.data = R_ADDR(dbmp->reginfo, mfp->pgcookie_off);
+        dbtp = &dbt;
+    }
+    if (is_pgin) {
+        if (dbmp->dbenv->pgin[mfp->ftype] != NULL &&
```

Comment:
Can ftype be something other than DB_BTREE? If not, turn this into an assert?

Reply:
Thanks Nirbhay.  This is actually badly broken (not in this code path) because ftype can be DB_FTYPE_SET (-1).  This still needs some work.  On the bright side, seeing visible performance improvement just from getting rid of this one extra lock in __dir_pg.